Public/Src/Cache/ContentStore/Vfs/VfsCasRunner.cs

```diff
@@ -0,0 +1,127 @@
+using BuildXL.Cache.ContentStore.Sessions;
+using BuildXL.Cache.ContentStore.Stores;
+
+namespace BuildXL.Cache.ContentStore.Vfs
+{
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Threading.Tasks;
+    using BuildXL.Cache.ContentStore.Interfaces.Extensions;
+    using BuildXL.Cache.ContentStore.Interfaces.Logging;
+    using BuildXL.Cache.ContentStore.Interfaces.Results;
+    using BuildXL.Cache.ContentStore.Interfaces.Stores;
+    using BuildXL.Cache.ContentStore.Interfaces.Tracing;
+    using BuildXL.Cache.ContentStore.Tracing;
+    using BuildXL.Cache.ContentStore.Tracing.Internal;
+    using BuildXL.Native.IO;
+    using static Placeholder;
+
+    /// <summary>
+    ///     Entry point of the application.
+    /// </summary>
+    public class VfsCasRunner
+    {
+        private Tracer Tracer { get; } = new Tracer(nameof(VfsCasRunner));
+
+        public async Task RunAsync(VfsCasConfiguration configuration)
+        {
+            // Create VFS root
+            using (var fileLog = new FileLog((configuration.RootPath / "Bvfs.log").Path))
+            using (var logger = new Logger(fileLog))
+            {
+                var fileSystem = new PassThroughFileSystem(logger);
+                var context = new OperationContext(new Context(logger));
+
+                // Map junctions into VFS root
+                foreach (var mount in configuration.VirtualizationMounts)
+                {
+                    CreateJunction(context, source: mount.Value, target: configuration.VfsRootPath / mount.Key);
+                }
+
+                var clientContentStore = new ServiceClientContentStore(
+                    logger,
+                    fileSystem,
+                    new ServiceClientContentStoreConfiguration(
+                        configuration.CacheName,
+                        new ServiceClientRpcConfiguration(
+                            configuration.BackingGrpcPort),
+                        scenario: configuration.Scenario));
+
+                // Client is startup/shutdown with wrapping VFS content store
+
+                using (var server = new LocalContentServer(
+                    fileSystem,
+                    logger,
+                    scenario: "bvfs" + configuration.ServerGrpcPort,
+                    path => new VirtualizedContentStore(clientContentStore),
+                    new LocalServerConfiguration(
+                        configuration.DataRootPath,
+                        new Dictionary<string, AbsolutePath>()
+                        {
+                        { configuration.CacheName, configuration.ServerRootPath }
+                        },
+                        configuration.ServerGrpcPort)))
+                {
+
+                    await server.StartupAsync(context).ThrowIfFailure();
+
+                    await WaitForTerminationAsync(context);
+
+                    await server.ShutdownAsync(context).ThrowIfFailure();

```

Comment:
Just curious, should we shutdown in a finally block?

Reply:
I don't think it makes sense to shutdown if you fail to startup.

---
In reply to: [281273745](https://github.com/Microsoft/BuildXL/pull/249#discussion_r281273745) [](ancestors = 281273745)